File: src/BFS.java
import java.util.*;

public class BFS {
    // This method takes in a 2d array of strings, a starting location, and an
    // ending location, and uses breadth first search to find the shortest path
    // between the two points. The path can only follow nodes marked "f" and
    // must avoid other nodes. It returns a list of coordinates which is the
    // path from start location to end location.
    public static List<int[]> shortestPath(char[][] grid, int[] start, int[] end) {
        // We will use a queue to keep track of which nodes we need to visit next
        // in our breadth first search.
        Queue<int[]> queue = new LinkedList<>();

        // We will also use a set to keep track of which nodes we have already
        // visited, so we don't visit them again and get stuck in a loop.
        Set<int[]> visited = new HashSet<>();

        // We will use a map to keep track of the previous node for each node
        // we visit, so we can reconstruct the path when we are finished.
        Map<int[], int[]> prev = new HashMap<>();

        // Add the starting node to the queue and mark it as visited
        queue.add(start);
        visited.add(start);

        // Perform the search
        while (!queue.isEmpty()) {
            // Get the next node to visit from the queue
            int[] node = queue.remove();

            // Check if we have reached the end node. If so, we can reconstruct
            // the path and return it.
            if (node[0] == end[0] && node[1] == end[1]) {
                List<int[]> path = new ArrayList<>();
                while (node != null) {
                    path.add(0, node);
                    node = prev.get(node);
                }
                return path;
            }

            // Otherwise, we need to visit the node's neighbors.
            for (int i = -1; i <= 1; i++) {
                for (int j = -1; j <= 1; j++) {
                    // We only want to visit the neighbors that are directly
                    // adjacent to the current node (not diagonally)
                    if (Math.abs(i) == Math.abs(j)) {
                        continue;
                    }

                    // Calculate the coordinates of the neighbor
                    int x = node[0] + i;
                    int y = node[1] + j;

                    // Make sure the coordinates are within the bounds of the grid
                    if (x < 0 || x >= grid.length || y < 0 || y >= grid[0].length) {
                        continue;
                    }

                    // Make sure we are only visiting nodes marked "f"
                    if (Character.compare(grid[x][y], 'f') == 0) {
                        continue;
                    }

                    // Create an array to represent the coordinates of the neighbor
                    int[] neighbor = {x, y};

                    // Make sure we haven't already visited this node
                    if (visited.contains(neighbor)) {
                        continue;
                    }

                    // Add the neighbor to the queue, mark it as visited, and
                    // record the current node
                    queue.add(neighbor);
                    visited.add(neighbor);
                    prev.put(neighbor, node);
                }
            }
        }

        // If we reach this point, it means we were unable to find a path from
        // the start node to the end node, so we return an empty list
        return Collections.emptyList();
    }

}

File: Simulator/Navigation/BFS.ts
import iTile from "../Room/iTile";
import Room from "../Room/Room";
import { Chair, Floor, Rug } from "../Room/Tiles";
import Coordinate from "./Coordinate";

const shortestPath = (room: Room, startPos: Coordinate, endPos: Coordinate): Coordinate[] => {
  // check if start and end path are the same
  if (startPos.x === endPos.x && startPos.y === endPos.y) {
    return [new Coordinate(endPos.x, endPos.y)];
  }

  let path: Coordinate[] = [];

  // check if startPos is valid
  if (!(room.RoomArray[startPos.y][startPos.x] instanceof Floor) && !(room.RoomArray[startPos.y][startPos.x] instanceof Rug) && !(room.RoomArray[startPos.y][startPos.x] instanceof Chair)) {
    console.error("No available path, start position invalid");
    return [];
  }
  // check if endPos is valid
  if (!(room.RoomArray[endPos.y][endPos.x] instanceof Floor || room.RoomArray[endPos.y][endPos.x] instanceof Rug || room.RoomArray[endPos.y][endPos.x] instanceof Chair)) {
    console.error("No available path, end position is not valid");
    return [];
  }

  const p = bfs(room, startPos, endPos);

  if (p.length === 0) {
    console.error("No available path");
  }

  // convert path into array of Coordinates
  for (const numberArr of p) {
    path.push(new Coordinate(numberArr[0], numberArr[1]));
  }
  return path;
};

function bfs(room: Room, startPos: Coordinate, endPos: Coordinate) {
  /* 
    Init 2d array 
    simplify roomArray into navigatable tiles and non-navigatable tiles
    floor & rug are navigatable, everything else is not

    convert it to a number array
    0 is walkable
    1 is not
    */

  const start = [startPos.x, startPos.y];
  const end = [endPos.x, endPos.y];

  const grid: number[][] = convertRoomArray(room.RoomArray);

  // // TODO COMMENT OUT, THIS IS FOR TESTING
  // // LOGGING grid
  // room.render()
  // let str = ""
  // for (const row of grid) {
  //     str += row + "\n"
  // }
  // console.log(str)
  // y,x
  // console.log(grid[5][2])

  // Simulate a queue by an array of coordinates
  // add to queue by doing queue = [...newCoord, ...queue]
  // or queue.unshift(newCoord)
  // remove from queue with queue.pop()

  let queue: number[][] = [];
  let visited = new Set<number[]>();
  let prev = new Map<number[], number[]>();

  queue.push(start);
  visited.add(start);

  // While the queue is not empty
  while (queue.length > 0) {
    // get the next node
    let node = queue.pop();
    // Typescript type check
    if (node === undefined) {
      throw new Error("node is undefined");
    }

    // Check if end position is reached, if so, we can reconstruct the path and return it
    if (node[0] === end[0] && node[1] === end[1]) {
      let path: number[][] = [];
      while (node !== null && node !== undefined) {
        path.unshift(node);
        node = prev.get(node);
      }
      return path;
    }

    // Otherwise, visit the node's neighbors
    for (let i = -1; i <= 1; i++) {
      for (let j = -1; j <= 1; j++) {
        // Directly adjacent to current node
        // Not diagonnaly, remove this if you want diagonally
        if (Math.abs(i) === Math.abs(j)) {
          continue;
        }

        // Calc coordinates of neighbor
        let x = node[0] + i;
        let y = node[1] + j;

        // Make sure coordinates are in bounds
        if (x < 0 || y < 0 || x >= grid.length || y >= grid[0].length) {
          continue;
        }

        // make sure only visiting viable nodes (0)
        if (grid[x][y] === 1) {
          continue;
        }

        // Create an array to represent the coordinates of the neighbor
        let neighbor: number[] = [x, y];
        // Make sure we haven't already visited this node
        if (visited.has(neighbor)) {
          continue;
        }

        // Add neighbor to queue, mark it as visited, record current node
        queue.unshift(neighbor);
        visited.add(neighbor);
        prev.set(neighbor, node);
      }
    }
  }
  // If we reach this point, unable to find path
  return [];
}

function convertRoomArray(roomArr: Array<Array<iTile>>): number[][] {
  let arr: number[][] = [];
  if (roomArr.length === 0) {
    throw new Error("This room is empty.");
  }

  for (let y = 0; y < roomArr.length; y++) {
    const row: number[] = [];
    for (let x = 0; x < roomArr[0].length; x++) {
      // TODO For now, sprites can walk through chairs
      if (roomArr[y][x] instanceof Floor || roomArr[y][x] instanceof Rug || roomArr[y][x] instanceof Chair) {
        row.push(0);
      } else {
        row.push(1);
      }
    }
    arr.push(row);
  }
  const newArr = transpose(arr);

  return newArr;
}

function transpose(matrix: number[][]) {
  const rows = matrix.length,
    cols = matrix[0].length;
  const grid = [];
  for (let j = 0; j < cols; j++) {
    grid[j] = Array(rows);
  }
  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      grid[j][i] = matrix[i][j];
    }
  }
  return grid;
}

export default shortestPath;
